bench: add "priority level" to the benchmark framework

furszy opened the pull request:
This is from today's meeting, a simple "priority level" for the benchmark framework.

Will allow us to run certain benchmarks while skip non-prioritized ones in `make check`.

By default, `bench_bitcoin` will run all the benchmarks. `make check`will only run the high priority ones,
and have marked all the existent benchmarks as "high priority" to retain the current behavior.

Could test it by modifying any benchmark priority to something different from "high", and
run `bench_bitcoin -priority-level=high` and/or `bench_bitcoin -priority-level=medium,low`
(the first command will skip the modified bench while the second one will include it).

Note: the second commit could be avoided by having a default arg value for the priority
level but.. an explicit set in every `BENCHMARK` macro call makes it less error-prone.

furszy committed:
bench: add "priority level" to the benchmark framework

Will allow us to run certain benchmarks while skip
non-prioritized ones in 'make check'.

martinus commented on src/bench/bench.h:
```diff
@@ -41,6 +41,16 @@ using ankerl::nanobench::Bench;
 
 typedef std::function<void(Bench&)> BenchFunction;
 
+enum BenchPriorityLevel
```
I'd use `enum class`

martinus replied:
think it would be nice if the enum would be some kind of bitmask instead. Then the entries act more like tags, and it would be possible to e.g. run all high and medium benchmarks by specifying both.

furszy replied:
yeah, much better.
pushed it.

martinus reviewed (commented):
quick code review ACK

luke-jr commented:
I think the default should be to run them all?

`make check` already overrides the number of iterations, and could easily limit it to just high priority.

furszy committed:
bench: surround main() execution with try/catch

so we have a cleaner exit on internal runtime errors.
e.g. an unknown priority level.

furszy committed:
bench: explicitly make all current benchmarks "high" priority

no-functional changes. Only have set the priority level explicitly
on every BENCHMARK macro call.

furszy committed:
build: only run high priority benchmarks in 'make check'

furszy commented:
Feedback tackled, thanks both

Changes:

* Moved priority level to be a bitfield. So combinations between different levels are allowed.
Now can pass `-priority-level=medium,low` or `-priority-level=all`.
* Moved default priority level to "all", and adapted `make check` to only run high priority ones.
* Surrounded bench `main()` with try/catch, mainly to have a cleaner exit in args
parsing errors (e.g. "unknown priority level").

martinus commented:
I like it, code review ACK  f95ce9b

stickies-v commented on src/bench/bench.cpp:
```diff
@@ -63,6 +85,11 @@ void benchmark::BenchRunner::RunAll(const Args& args)
     std::vector<ankerl::nanobench::Result> benchmarkResults;
     for (const auto& p : benchmarks()) {
+
+        if (!(p.second.second & args.priority_level)) {
```
Will expand the diff a bit, but wdyt about using structured bindings for iterating over `benchmarks()` and unpacking the second pair (as you can't do nested structured bindings)? `p.second.second` could benefit from some improved readability.

furszy replied:
sure, give me a second.

stickies-v reviewed (commented):
Concept ACK - looks like a good improvement to allow `make check` not become too slow and still allow flexibility as to which benchmarks we want to add to the test suite.

furszy commented:
Updated per feedback.

Pairs unpacked, small [diff](https://github.com/bitcoin/bitcoin/compare/74ac7142d7ce3cd6012c79f38cb540ba3650a999..23b6a7ab5b4003e11303af3f779803eb79d0bc58)

stickies-v commented on src/bench/bench.cpp:
```diff
@@ -41,15 +41,37 @@ void GenerateTemplateResults(const std::vector<ankerl::nanobench::Result>& bench
+    for (const auto& it : map_priority_level) {
+        if (it.second == level) return it.first;
+    }
+    throw std::runtime_error(strprintf("Unknown priority level %d", level));
```
I'm not sure why this compiles because I think you can't pass a `PriorityLevel` like that:

```suggestion
    throw std::runtime_error(strprintf("Unknown priority level %u", static_cast<uint8_t>(level)));
```

Also, I don't think this is a runtime error because it can only be hit when `map_priority_level` is out of sync with `PriorityLevel`?

furszy replied:
Better `PriorityToString(level)` instead. The number isn't so informative.

stickies-v commented on src/bench/bench_bitcoin.cpp:
```diff
@@ -30,6 +30,9 @@ static void SetupBenchArgs(ArgsManager& argsman)
     argsman.AddArg("-sanity-check", "Run benchmarks for only one iteration", ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
+    argsman.AddArg("-priority-level=<l1,l2,l3>", strprintf("Run benchmarks of a certain priority level (%s | %s | %s), default: %s", PriorityToString(benchmark::PriorityLevel::LOW),
+                                                           PriorityToString(benchmark::PriorityLevel::MEDIUM), PriorityToString(benchmark::PriorityLevel::HIGH),
+                                                           PriorityToString(benchmark::PriorityLevel::ALL)), ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
```
```suggestion
                                                           PriorityToString(benchmark::DEFAULT_PRIORITY_LEVEL)), ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
```

stickies-v reviewed (commented):
Went through most of the code now, a few more mostly minor comments.

nit: From a readability pov, I'm a bit uneasy with `priority_level` sometimes being a `uint8_t` and sometimes a `PriorityLevel`. Not very intuitive, but I don't have a clear solution atm. (I do like the bitmask approach).

furszy commented on src/bench/bench_bitcoin.cpp:
```diff
@@ -30,6 +30,9 @@ static void SetupBenchArgs(ArgsManager& argsman)
     argsman.AddArg("-output-csv=<output.csv>", "Generate CSV file with the most important benchmark results", ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
     argsman.AddArg("-output-json=<output.json>", "Generate JSON file with all benchmark results", ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
     argsman.AddArg("-sanity-check", "Run benchmarks for only one iteration", ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
+    argsman.AddArg("-priority-level=<l1,l2,l3>", strprintf("Run benchmarks of a certain priority level (%s | %s | %s), default: %s", PriorityToString(benchmark::PriorityLevel::LOW),
```
I would try to not pollute the flag name with specific parameters. So that it does not depend on the format function output (same behavior as every other flag here and `init.cpp`).

furszy commented on src/bench/bench.h:
```diff
@@ -41,6 +41,20 @@ using ankerl::nanobench::Bench;
+/** Priority level default value */
+static const PriorityLevel DEFAULT_PRIORITY_LEVEL = PriorityLevel::ALL;
+
+uint8_t ToPriorityLevel(const std::string& str);
+std::string PriorityToString(PriorityLevel level);
```
sure

furszy commented on src/bench/bench_bitcoin.cpp:
```diff
@@ -45,6 +48,17 @@ static std::vector<double> parseAsymptote(const std::string& str) {
 }
 
+static uint8_t parsePriorityLevel(const std::string& str) {
+    std::stringstream ss(str);
+    uint8_t levels = 0;
+    while (ss.good()) {
+        std::string level;
+        std::getline(ss, level, ',');
+        levels |= benchmark::ToPriorityLevel(level);
+    }
+    return levels;
```
Sure. Same function can be used inside `parseAsymptote`.

furszy commented on src/bench/bench_bitcoin.cpp:
```diff
@@ -30,6 +30,9 @@ static void SetupBenchArgs(ArgsManager& argsman)
     argsman.AddArg("-output-csv=<output.csv>", "Generate CSV file with the most important benchmark results", ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
     argsman.AddArg("-output-json=<output.json>", "Generate JSON file with all benchmark results", ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
     argsman.AddArg("-sanity-check", "Run benchmarks for only one iteration", ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
+    argsman.AddArg("-priority-level=<l1,l2,l3>", strprintf("Run benchmarks of a certain priority level (%s | %s | %s), default: %s", PriorityToString(benchmark::PriorityLevel::LOW),
```
Done

furszy commented:
Updated per feedback, thanks @stickies-v  👍🏼.

Changes [diff](https://github.com/bitcoin/bitcoin/compare/23b6a7ab5b4003e11303af3f779803eb79d0bc58..6e95eaede1c7fe477836ce0cb1680aa59ecec51d):
* Renamed `ToPriorityLevel` to `StringToPriority`.
* Moved `PriorityToString` runtime_error to assertion.
* Adjusted the new flag help message description.
* Made `parsePriorityLevel` use `SplitString` function.

stickies-v commented on src/bench/bench.cpp:
```diff
@@ -41,15 +41,37 @@ void GenerateTemplateResults(const std::vector<ankerl::nanobench::Result>& bench
+    for (const auto& [name, level] : map_priority_level) {
+        if (level == in_level) return name;
+    }
+    assert(false); // out-of-sync `map_priority_level`
```
requires `#include <cassert>`

You could also use `assert(!"<message>")` as in: https://github.com/bitcoin/bitcoin/blob/9fcdb9f3a044330d3d7515fa35709102c98534d2/src/net_processing.cpp#L2145

stickies-v commented on src/bench/bench.cpp:
```diff
@@ -41,15 +41,37 @@ void GenerateTemplateResults(const std::vector<ankerl::nanobench::Result>& bench
+    return it->second;
+}
+
+std::string benchmark::PriorityToString(PriorityLevel in_level)
```
Currently, there's a bit of inconsistency in the namespace specification (`benchmark::PriorityLevel` in other places). What do you think about putting all the `benchmark` implementations here in `namespace benchmark` and remove all the `benchmark::` specifiers? This makes the code more concise and explicit, and mirrors the header file. (thank you @hebasto for the idea).

furszy replied:
sure, I thought about it as well but didn't want to expand this further.
Now that it's not just me, will add a small commit for it at the beginning :).

stickies-v reviewed (commented):
Code Review ACK 6e95eaede

Will do some testing shortly.

furszy committed:
bench: place benchmark implementation inside benchmark namespace

furszy commented on src/bench/bench_bitcoin.cpp:
```diff
@@ -45,6 +48,14 @@ static std::vector<double> parseAsymptote(const std::string& str) {
+static uint8_t parsePriorityLevel(const std::string& str) {
+    uint8_t levels = 0;
+    for (const auto& level: SplitString(str, ',')) {
+        levels |= static_cast<uint8_t>(benchmark::StringToPriority(level));
```
That is interesting. Wasn't expecting to be able to store not declared enum combinations inside the enum variable.
So.. same as happens with the `NetPermissionFlags`,  could remove every static_cast and do something like:
```c++
PriorityLevel levels{0};
for (const auto& level: SplitString(str, ',')) {
   levels = levels | benchmark::StringToPriority(level);
}
return levels;
```

Which.. can make the software crash if I pass a combination of tags that isn't declared (e.g. "medium, low") and call `PriorityToString(levels)` (at the assertion point).

furszy commented:
Feedback tackled, thanks both.

Changes ([diff](https://github.com/bitcoin/bitcoin/compare/6e95eaede1c7fe477836ce0cb1680aa59ecec51d..9c3ab931fbfbec5691a44de5f21138654db9f2fd)):

- Placed the benchmark implementation inside benchmark namespace.
- Added missing `cassert` include.
- Removed static_assert by defining bitwise operators on the enum class.

stickies-v commented on src/bench/bench.cpp:
```diff
@@ -41,18 +42,44 @@ void GenerateTemplateResults(const std::vector<ankerl::nanobench::Result>& bench
-benchmark::BenchRunner::BenchmarkMap& benchmark::BenchRunner::benchmarks()
+namespace benchmark {
+
+std::map<std::string, PriorityLevel> map_priority_level = {
```
Since the previous force push, we now can't pass `-priority-level=all` anymore, I don't think that's intentional? And if it was, I don't think I agree with that `-priority-level=low,medium,high` is cumbersome and will silently be incomplete when adding new levels.

furszy replied:
> Since the previous force push, we now can't pass -priority-level=all anymore, I don't think that's intentional?

Yes, to avoid tags clashing problems. e.g. when you call `PriorityToString` with any value, the returned string was the actual priority and, as it matches with the rest of the tags, the "all" tag.

But.. will re-add the "all" shortcut differently now.

And.. you know, I actually forgot to go further here, should had removed the `PriorityLevel::ALL` and replace it by a plain 0xff.
We shouldn't allow benchmarks to be set under the `PriorityLevel::ALL` tag as you did above (the `BENCHMARK(AddrManAddThenGood, benchmark::PriorityLevel::ALL)`).

> And if it was, I don't think I agree with that -priority-level=low,medium,high is cumbersome and will silently be incomplete when adding new levels.

With the latest change (0xff inclusion), all levels will be automatically added there.

stickies-v reviewed (commented):
Did some first testing already, everything seems to work as expected (except for `-priority-level=all` not working anymore as commented).

I'm not sure we want to do something about it, but currently we can also specify a benchmark as `benchmark::PriorityLevel::ALL`, in which case it would get run for any `-priority-level`.

E.g. with
```cpp
BENCHMARK(AddrManAdd, benchmark::PriorityLevel::LOW);
BENCHMARK(AddrManSelect, benchmark::PriorityLevel::MEDIUM);
BENCHMARK(AddrManGetAddr, benchmark::PriorityLevel::HIGH);
BENCHMARK(AddrManAddThenGood, benchmark::PriorityLevel::ALL);
```

For `./bench_bitcoin --priority-level=low` we get:
```
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|       19,047,208.00 |               52.50 |    1.8% |      0.21 | `AddrManAdd`
|       48,141,333.00 |               20.77 |    1.8% |      0.53 | `AddrManAddThenGood`
```

furszy commented on src/bench/bench.cpp:
```diff
@@ -41,18 +42,44 @@ void GenerateTemplateResults(const std::vector<ankerl::nanobench::Result>& bench
+    for (const auto& [name, level] : map_priority_level) {
+        if (level & in_level) tags += name + ",";
+    }
+    tags.pop_back(); // remove the last comma
+    assert(!tags.empty()); // out-of-sync `map_priority_level`
```
yeah good

amovfx commented on src/bench/bench.cpp:
```diff
@@ -41,18 +42,46 @@ void GenerateTemplateResults(const std::vector<ankerl::nanobench::Result>& bench
+
+PriorityLevel StringToPriority(const std::string& str)
+{
+    if (str == "all") return (PriorityLevel)0xff; // "all" shortcut
+    auto it = map_priority_level.find(str);
+    if (it == map_priority_level.end()) throw std::runtime_error(strprintf("Unknown priority level %s", str));
```
Is there a valid concern that the string could be entered in as multicase? Like Low, Medium, High, All?
Maybe this would be better suited in parsePriorityLevel?
```suggestion
    std::transform(str.begin(), str.end(), str.begin(), ::tolower);
    if (str == "all") return (PriorityLevel)0xff; // "all" shortcut
    auto it = map_priority_level.find(str);
    if (it == map_priority_level.end()) throw std::runtime_error(strprintf("Unknown priority level %s", str));
```

furszy replied:
Something like that would be better suited for the general `ArgsManager::ParseParameters` function. Not only for this new flag.

But I'm not so sure about it. Init flags are keys, and keys shouldn't have different values. Should be a 1-1 map.

amovfx replied:
Makes sense, I was thinking the same thing last night. ArgsManager should perhapse enforce case if that is a good idea. Thanks.

amovfx reviewed (commented):
Reviewed for pr review club.
I was surprised to see all the benchmarks that were updated got benchmark::PriorityLevel::HIGH.

furszy commented:
> Reviewed for pr review club.
> I was surprised to see all the benchmarks that were updated got benchmark::PriorityLevel::HIGH.

Why? It's stated in the PR description:

By default, bench_bitcoin will run all the benchmarks. `make check` will only run the high priority ones,
and have marked all the existent benchmarks as "high priority" to retain the current behavior.

The goal is to not introduce a behavior change here, only add the new functionality. Then discuss specific cases on parallel independent PRs.

amovfx commented:
@furszy 
Sorry, I need to read better.

hernanmarino reviewed (commented):
code review ACK . Thoroughly tested changing priorities, everything works fine. I like the approach of the bit mapping for priorities

pablomartin4btc reviewed (commented):
Tested ACK,

Compiled this pr, changed a benchmark priority and recompiled only the [bitcoin_bench](https://github.com/bitcoin/bitcoin/blob/master/doc/benchmarking.md) (`make -C src bitcoin_bench`) playing with the different priority levels (low, medium, high, all, none) and some combinations, it works as expected.

All existing benchmarks are set with "high priority" to match the current behaviour where all benchmarks run by default on "`make check`", so there's no impact merging the change.

I understand @stickies-v's point of view about the inconsistecy between the priority level enum and the bitvector, in terms of their original purpose, as the enum specifies items from a limited set, the uint8_t mask/ mapping would represent the combination/ agreggation of all of them. I don't have a strong position on how to unlink the uint8_t use case as on the other side it looks a very good approach.

kouloumos commented on src/bench/bench_bitcoin.cpp:
```diff
@@ -30,6 +30,9 @@ static void SetupBenchArgs(ArgsManager& argsman)
     argsman.AddArg("-output-csv=<output.csv>", "Generate CSV file with the most important benchmark results", ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
     argsman.AddArg("-output-json=<output.json>", "Generate JSON file with all benchmark results", ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
     argsman.AddArg("-sanity-check", "Run benchmarks for only one iteration", ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
+    argsman.AddArg("-priority-level=<l1,l2,l3>", strprintf("Run benchmarks of one or multiple priority level(s) (%s | %s | %s), default: '%s'", PriorityToString(benchmark::PriorityLevel::LOW),
+                                                           PriorityToString(benchmark::PriorityLevel::MEDIUM), PriorityToString(benchmark::PriorityLevel::HIGH),
+                                                           PriorityToString(benchmark::DEFAULT_PRIORITY_LEVEL)), ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
```
I am not sure that I see the benefit of the `PriorityToString` function, I believe that it adds extra circular transformation complexity (priority -> string -> priority) without a solid benefit.

In case of a change to priorities (adding/removing options), we will still need to make a change here. Maybe a function that uses `map_priority_level` to concatenate all available options could be a better fit. With this changed what's left for `PriorityToString` is its usage with `DEFAULT_PRIORITY_LEVEL` which can be erased by using string instead of priority:`DEFAULT_PRIORITY_LEVEL{"all"}`.
(afc9c1bb544dfc5418102dd2e5f9f2819b24b885)

furszy replied:
sounds good, pushed inside 085e88ef.

kouloumos reviewed (commented):
ACK cbc077e77a5d0ba8ae11b1f5d89480c997fdef55 minus my comments.

Do we have a heuristic on the expected behavior when passing an arg without options? e.g `bench_bitcoin -priority-level`? Although I don't consider it an issue, I observed that there are inconsistencies (which can be traced back to how each arg is parsing the empty string `""`):
-  `-asymptote`, `-min-time`, `-output-csv`, `-output-json`: behave like the arg was never given.
- `filter`: filters with the empty string "", therefore no benchmarks are executed.
- `-priority-level`: Errors with unknown priority level

furszy commented:
> Do we have a heuristic on the expected behavior when passing an arg without options? e.g `bench_bitcoin -priority-level`? Although I don't consider it an issue, I observed that there are inconsistencies (which can be traced back to how each arg is parsing the empty string `""`):
> 
> * `-asymptote`, `-min-time`, `-output-csv`, `-output-json`: behave like the arg was never given.
> * `filter`: filters with the empty string "", therefore no benchmarks are executed.
> * `-priority-level`: Errors with unknown priority level

Unlike the other args, the `-priority-level` inputted string need to match against one of the existent levels.

The `filter` arg behavior make sense because it's a pattern matching string. e.g. if you provide an empty string, then nothing will match against it. So no benchmarks will be executed.

For the others, probably it's not a bad idea to return an error (or print to console) when an empty flag is passed. So the user knows that the arg is wrong.

furszy commented on src/bench/bench.h:
```diff
@@ -49,22 +72,24 @@ struct Args {
     fs::path output_csv;
     fs::path output_json;
     std::string regex_filter;
+    PriorityLevel priority_level{DEFAULT_PRIORITY_LEVEL};
```
done

furszy commented:
Updated per @kouloumos feedback.

Have removed `PriorityToString()`, and replaced it by `ListPriorities()`.

stickies-v commented on src/bench/bench_bitcoin.cpp:
```diff
@@ -18,6 +18,8 @@
 static const char* DEFAULT_BENCH_FILTER = ".*";
 static constexpr int64_t DEFAULT_MIN_TIME_MS{10};
+/** Priority level default value, run "all" priority levels */
+static const std::string DEFAULT_PRIORITY_LEVEL{"all"};
```
Not a huge concern, but I'm not terribly excited about this string reference. Would be nicer imo to have the default be a `uint8_t` so it can't go out of sync, but that would require the `PriorityToString` fn again which may be overkill.

furszy replied:
yeah. Same feeling here. But.. would say to leave it as is for now, otherwise it would mean to re-introduce the `PriorityToString` function only to stringify this priority level.

stickies-v commented on src/bench/bench.cpp:
```diff
@@ -41,18 +42,44 @@ void GenerateTemplateResults(const std::vector<ankerl::nanobench::Result>& bench
+};
+
+std::string ListPriorities()
+{
+    std::string tags;
+    for (const auto& [name, level] : map_priority_level) {
+        tags += name + ",";
+    }
+    tags.pop_back(); // remove the last comma
+    return tags;
+}
```
With `#include <util/string.h>`:
```suggestion
{
    return Join(map_priority, ',', [](const auto& entry){  return entry.first; });
}
```

stickies-v replied:
Not a huge fan of how this sorts alphabetically by label instead of by increasing priority, but... not a very straightforward around that it seems. Just raising in case you have ideas.

furszy replied:
> Not a huge fan of how this sorts alphabetically by label instead of by increasing priority, but... not a very straightforward way around that it seems. Just raising in case you have ideas.

Could add a custom comparator. One sec.

stickies-v reviewed (commented):
As discussed before, I'm not a huge fan of using `PriorityLevel` both as an enum (corresponding with a single, enumerated flag to indicate the level of each benchmark) and the bitvector representing a (potential) aggregation of multiple levels. I think this makes the code less clear, and I'd prefer an enum to be used as an actual enum instead of some hybrid form.

What do you think about this diff?
- use a `uint8_t` for the bitvector
- use an `enum` instead of `enum class` so we can implicitly convert to its underlying type. I know this contradicts [an earlier comment](https://github.com/bitcoin/bitcoin/pull/26158#discussion_r979171783) but I think it makes sense here, and the code is also significantly different from the one where the comment was made
- no need for a manual `if "all"` check, generalizes a bit better

In my opinion it makes the intent of the code more clear, and also reduces the LoC required for this PR.

Note: I think after this diff, some variables should be renamed to drop "level" in their name, but I did not do that here to keep the diff clear.

<details>
<summary>git diff</summary>

```diff
diff --git a/src/bench/bench.cpp b/src/bench/bench.cpp
index d2749be1c..37785759e 100644
--- a/src/bench/bench.cpp
+++ b/src/bench/bench.cpp
@@ -45,10 +45,11 @@ void GenerateTemplateResults(const std::vector<ankerl::nanobench::Result>& bench
 
 namespace benchmark {
 
-std::map<std::string, PriorityLevel> map_priority_level = {
+std::map<std::string, uint8_t> map_priority_level = {
     std::make_pair("high", PriorityLevel::HIGH),
     std::make_pair("medium", PriorityLevel::MEDIUM),
-    std::make_pair("low", PriorityLevel::LOW)
+    std::make_pair("low", PriorityLevel::LOW),
+    std::make_pair("all", 0xff)
 };
 
 std::string ListPriorities()
@@ -56,9 +57,8 @@ std::string ListPriorities()
     return Join(map_priority_level, ',', [](const auto& entry){  return entry.first; });
 }
 
-PriorityLevel StringToPriority(const std::string& str)
+uint8_t StringToPriority(const std::string& str)
 {
-    if (str == "all") return (PriorityLevel)0xff; // "all" shortcut
     auto it = map_priority_level.find(str);
     if (it == map_priority_level.end()) throw std::runtime_error(strprintf("Unknown priority level %s", str));
     return it->second;
diff --git a/src/bench/bench.h b/src/bench/bench.h
index 295d20bf0..08c682d97 100644
--- a/src/bench/bench.h
+++ b/src/bench/bench.h
@@ -41,25 +41,15 @@ using ankerl::nanobench::Bench;
 
 typedef std::function<void(Bench&)> BenchFunction;
 
-enum class PriorityLevel : uint8_t
+enum PriorityLevel : uint8_t
 {
     LOW = 1 << 0,
     MEDIUM = 1 << 1,
     HIGH = 1 << 2,
 };
-static inline constexpr PriorityLevel operator|(PriorityLevel a, PriorityLevel b)
-{
-    using t = typename std::underlying_type<PriorityLevel>::type;
-    return static_cast<PriorityLevel>(static_cast<t>(a) | static_cast<t>(b));
-}
-static inline constexpr bool operator&(PriorityLevel tags, PriorityLevel tag)
-{
-    using t = typename std::underlying_type<PriorityLevel>::type;
-    return static_cast<t>(tags) & static_cast<t>(tag);
-}
 
 std::string ListPriorities();
-PriorityLevel StringToPriority(const std::string& str);
+uint8_t StringToPriority(const std::string& str);
 
 struct Args {
     bool is_list_only;
@@ -69,7 +59,7 @@ struct Args {
     fs::path output_csv;
     fs::path output_json;
     std::string regex_filter;
-    PriorityLevel priority_level;
+    uint8_t priority_level;
 };
 
 class BenchRunner
diff --git a/src/bench/bench_bitcoin.cpp b/src/bench/bench_bitcoin.cpp
index 54d5d67fa..2a1376aa9 100644
--- a/src/bench/bench_bitcoin.cpp
+++ b/src/bench/bench_bitcoin.cpp
@@ -49,8 +49,8 @@ static std::vector<double> parseAsymptote(const std::string& str) {
     return numbers;
 }
 
-static benchmark::PriorityLevel parsePriorityLevel(const std::string& str) {
-    benchmark::PriorityLevel levels{0};
+static uint8_t parsePriorityLevel(const std::string& str) {
+    uint8_t levels{0};
     for (const auto& level: SplitString(str, ',')) {
         levels = levels | benchmark::StringToPriority(level);
     }
 ```
</details>

furszy commented:
> As discussed before, I'm not a huge fan of using PriorityLevel both as an enum (corresponding with a single, enumerated flag to indicate the level of each benchmark) and the bitvector representing a (potential) aggregation of multiple levels. I think this makes the code less clear, and I'd prefer an enum to be used as an actual enum instead of some hybrid form.

Yeah, I wasn't a huge fan of it neither. Have applied all the suggestions, thanks @stickies-v :).

Plus, as said in https://github.com/bitcoin/bitcoin/pull/26158#discussion_r997335617, added a custom comparator to order the `ListPriorities()` outcome (squashed inside 3a09b15).

stickies-v commented on src/bench/bench.cpp:
```diff
@@ -41,18 +42,47 @@ void GenerateTemplateResults(const std::vector<ankerl::nanobench::Result>& bench
+std::string ListPriorities()
 {
-    static std::map<std::string, BenchFunction> benchmarks_map;
+    std::set<std::pair<std::string, uint8_t>, CompPriorities> vec(map_priority_level.begin(), map_priority_level.end());
```
Since we're only using it in this function, I think I'd prefer a local lambda over a struct. Easier to read imo:

<details>
<summary>git diff</summary>

```diff
diff --git a/src/bench/bench.cpp b/src/bench/bench.cpp
index 13d3c203d..b9a8e85c4 100644
--- a/src/bench/bench.cpp
+++ b/src/bench/bench.cpp
@@ -51,17 +51,12 @@ std::map<std::string, uint8_t> map_priority_level = {
     std::make_pair("all", 0xff)
 };
 
-struct CompPriorities {
-    template <typename T>
-    bool operator()(const T& l, const T& r) const {
-        return l.second < r.second;
-    }
-};
-
 std::string ListPriorities()
 {
-    std::set<std::pair<std::string, uint8_t>, CompPriorities> vec(map_priority_level.begin(), map_priority_level.end());
-    return Join(vec, ',', [](const auto& entry){ return entry.first; });
+    using item_t = std::pair<std::string, uint8_t>;
+    auto sort_by_priority = [](item_t a, item_t b){ return a.second < b.second; };
+    std::set<item_t, decltype(sort_by_priority)> sorted_priorities(map_priority_level.begin(), map_priority_level.end(), sort_by_priority);
+    return Join(sorted_priorities, ',', [](const auto& entry){ return entry.first; });
 }
 
 uint8_t StringToPriority(const std::string& str)
```
</details>

If not, I think naming could be improved, e.g. `sorted_priorities` instead of `vec` (it's not a vector) and `SortByPriority` instead of `CompPriorities`?

stickies-v commented on src/bench/bench.cpp:
```diff
@@ -41,18 +42,47 @@ void GenerateTemplateResults(const std::vector<ankerl::nanobench::Result>& bench
+    }
+};
+
+std::string ListPriorities()
```
```suggestion
// List priority labels, comma-separated and sorted by increasing priority
std::string ListPriorities()
```

stickies-v commented on src/bench/bench.cpp:
```diff
@@ -62,33 +92,39 @@ void benchmark::BenchRunner::RunAll(const Args& args)
     std::vector<ankerl::nanobench::Result> benchmarkResults;
-    for (const auto& p : benchmarks()) {
-        if (!std::regex_match(p.first, baseMatch, reFilter)) {
+    for (const auto& [name, bench_func] : benchmarks()) {
```
nit: since this unpacking touches a few lines, I think this would benefit from being split out into a separate refactor commit to make that more clear?

furszy replied:
I think that splitting it would be too much for such a small change.

stickies-v replied:
Okay, I don't agree but not that important so you can mark as resolved.

stickies-v commented on src/bench/bench.h:
```diff
@@ -49,22 +59,24 @@ struct Args {
     fs::path output_csv;
     fs::path output_json;
     std::string regex_filter;
+    uint8_t priority_level;
```
nit: I think `priority` better highlights the difference between the aggregated priority and the individual priority level
```suggestion
    uint8_t priority;
```

stickies-v commented on src/bench/bench_bitcoin.cpp:
```diff
@@ -18,6 +18,8 @@
 static const char* DEFAULT_BENCH_FILTER = ".*";
 static constexpr int64_t DEFAULT_MIN_TIME_MS{10};
+/** Priority level default value, run "all" priority levels */
+static const std::string DEFAULT_PRIORITY_LEVEL{"all"};
```
nit: I think `priority` better highlights the difference between the aggregated priority and the individual priority level
```suggestion
static const std::string DEFAULT_PRIORITY{"all"};
```

stickies-v commented on src/bench/bench.cpp:
```diff
@@ -41,18 +42,47 @@ void GenerateTemplateResults(const std::vector<ankerl::nanobench::Result>& bench
-benchmark::BenchRunner::BenchmarkMap& benchmark::BenchRunner::benchmarks()
+namespace benchmark {
+
+std::map<std::string, uint8_t> map_priority_level = {
```
nit: I think `priority` better highlights the difference between the aggregated priority and the individual priority level
```suggestion
std::map<std::string, uint8_t> map_label_priority = {
```

stickies-v commented on src/bench/bench.cpp:
```diff
@@ -41,18 +42,47 @@ void GenerateTemplateResults(const std::vector<ankerl::nanobench::Result>& bench
+    }
+};
+
+std::string ListPriorities()
```
What do you think about returning a sorted `std::vector<std::string>` here to make it more reusable? The `Join()` can easily be applied directly in `bench_bitcoin.cpp`.

furszy replied:
better to add it when we get an actual use case for it. For now, I think that is fine as is.

stickies-v replied:
Okay, I don't agree but not that important so you can mark as resolved.

stickies-v reviewed (approved):
ACK bacb1aef9

A couple more suggestions and nits but none of them blocking and could be done in follow-up too if preferred.

furszy commented:
Thanks @stickies-v for the feedback!, applied most of them.

Pure styling [diff](https://github.com/bitcoin/bitcoin/compare/bacb1aef9c18d87cf6668f6ad057dbef4e7d9720..06664b8ef0c5eb85c63adb573724af0e3541933f), no behavior change.

Should be ready to go now.

stickies-v commented on src/bench/bench.cpp:
```diff
@@ -62,33 +88,39 @@ void benchmark::BenchRunner::RunAll(const Args& args)
-    for (const auto& p : benchmarks()) {
-        if (!std::regex_match(p.first, baseMatch, reFilter)) {
+    for (const auto& [name, bench_func] : benchmarks()) {
+        const auto& [func, priority] = bench_func;
```
nit: I think you renamed a bit too much - this one actually is a `priority_level`. My suggestion was to help distinguish between `PriorityLevel` and a `uint8_t`, not to just drop the "level" part everywhere. Same thing [here](https://github.com/bitcoin/bitcoin/compare/bacb1aef9c18d87cf6668f6ad057dbef4e7d9720..06664b8ef0c5eb85c63adb573724af0e3541933f#diff-6ab99df740dfacb06eb9c5897231240c2b7244d8c5d6e9de31b6912f9b4a5a69R67-R68) and [here](https://github.com/bitcoin/bitcoin/compare/bacb1aef9c18d87cf6668f6ad057dbef4e7d9720..06664b8ef0c5eb85c63adb573724af0e3541933f#diff-6ab99df740dfacb06eb9c5897231240c2b7244d8c5d6e9de31b6912f9b4a5a69R78-R81)

furszy replied:
pushed it.

stickies-v reviewed (approved):
re-ACK 010f53e

achow101 commented:
ACK c2f53f987415df8f41b22967e03d02e19ed5f46c

MarcoFalke commented:
Is there a guideline on how to decide the priority level? I'd presume this is subjective, or is it simply a performance question? That is, slow benchmarks have low priority?

furszy commented:
> Is there a guideline on how to decide the priority level? I'd presume this is subjective, or is it simply a performance question? That is, slow benchmarks have low priority?

Yeah ok. So far, the turning point here seems to be introduction of benchmarks who measure/test how certain flows behave on a large scale (context https://github.com/bitcoin/bitcoin/pull/25685#discussion_r975424674). Which, usually, require a somewhat slow context setup (e.g. fill-up the wallet with 20k utxo to benchmark the tx creation process, coin selection, etc) which goes outside of the scope of the "run all benchs" sanity check that we are currently doing in `make check`.

furszy commented:
> Ok, so this is similar to what the functional tests call "extended"? I am asking because right now it looks like this doesn't change the "priority" of any benchmark, and if one is changed we should have at least one CI task that runs all benchmarks, similar to how there is one that runs all functional tests.

Yeah, it's similar to the "extended" concept there. The goal is to not only be limited to have fast and small benchmarks due the `make check` duration concerns. Some big issues aren't problems at all in small scale.

And yeah, this PR does not change the current behavior in any way. Only introduces the filtering by priority functionality and, by default, makes `make check` only run the high priority ones. Then #25685 introduces two "low priority" benchmarks.

absolutely, a CI task to run them all would be nice, yeah.

furszy commented:
Side note, to make the priority level decision process simpler, could also remove the `medium` level. So we only have two options, `high` or `low` priority.

kouloumos commented:
> Side note, to make the priority level decision process simpler, could also remove the `medium` level. So we only have two options, `high` or `low` priority.

I was also thinking about that decision process where someone is deciding the priority level of his new benchmark. How to differentiate between "medium" and 'low"? Do you have any scenarios/thoughts for that?

Instead of 2 priorities, maybe we could go with the "extended" notion that functional tests use?

furszy commented:
> I was also thinking about that decision process where someone is deciding the priority level of his new benchmark. How to differentiate between "medium" and 'low"? Do you have any scenarios/thoughts for that?

I think that it depends on the benchmarked area mostly. E.g. If we are benchmarking a process from the validation code (like the duplicate-inputs benchmark), it should have higher priority than a benchmark that tests the `ListTransactions` RPC-only function.
Another example would be a bug in the wallet that slows down the block scanning process on big wallets by 10x. Which, in my view, should have lower priority than the duplicate-inputs bench (as it cannot be run in `make check` due the slow context setup requirements, because it might need to generate a 100k blocks chain), but higher than the `ListTransactions` one.

> Instead of 2 priorities, maybe we could go with the "extended" notion that functional tests use?

Could go with that terminology as well, yeah. But, I think that, if we define which areas, problems and time constrains are more suited for each priority level, then the different levels approach make more sense.

theStack commented on src/bench/bench.cpp:
```diff
@@ -43,15 +44,38 @@ void GenerateTemplateResults(const std::vector<ankerl::nanobench::Result>& bench
 
+// map a label to one or multiple priority levels
+std::map<std::string, uint8_t> map_label_priority = {
+    std::make_pair("high", PriorityLevel::HIGH),
+    std::make_pair("medium", PriorityLevel::MEDIUM),
+    std::make_pair("low", PriorityLevel::LOW),
+    std::make_pair("all", 0xff)
```
stylistic nit (feel free to ignore):
```suggestion
    {"high", PriorityLevel::HIGH},
    {"medium", PriorityLevel::MEDIUM},
    {"low", PriorityLevel::LOW},
    {"all", 0xff}
```

theStack commented on src/bench/bench_bitcoin.cpp:
```diff
@@ -45,6 +49,14 @@ static std::vector<double> parseAsymptote(const std::string& str) {
     return numbers;
 }
 
+static uint8_t parsePriorityLevel(const std::string& str) {
```
nit: I think we use `lowerCamelCase` only for interface method names (as per developer guidelines).
```suggestion
static uint8_t ParsePriorityLevel(const std::string& str) {
```

theStack reviewed (approved):
Code-review ACK 010f53e7da9602dc8c92623d4839b5e4c9a813fc ⛰️ 

Left two small nits below, but feel free to ignore. Happy to also re-review if you decide to change the priorities numbers/naming ("low"/"high" or "normal"/"extended" or whatever).

furszy commented:
Updated per feedback. Thanks everyone :).

Have removed the `medium` priority level for now, mainly to minimize future doubts/discussions around it. It could easily be re-added in the future (I left a blank space for it) when more benchmarks like the ones that described in https://github.com/bitcoin/bitcoin/pull/26158#issuecomment-1284189169 are introduced.

Tiny [diff](https://github.com/bitcoin/bitcoin/compare/010f53e7da9602dc8c92623d4839b5e4c9a813fc..3e9d0bea8deb61596c91ead997e9db83f5b0ff68).

Ready to go.

theStack reviewed (approved):
re-ACK 3e9d0bea8deb61596c91ead997e9db83f5b0ff68

stickies-v reviewed (approved):
re-ACK https://github.com/bitcoin/bitcoin/commit/3e9d0bea8deb61596c91ead997e9db83f5b0ff68

kouloumos commented:
re-ACK 3e9d0bea8deb61596c91ead997e9db83f5b0ff68

achow101 commented:
ACK 3e9d0bea8deb61596c91ead997e9db83f5b0ff68

hebasto commented on src/Makefile.test.include:
```diff
@@ -373,8 +373,8 @@ endif
-	@echo "Running bench/bench_bitcoin (one iteration sanity check)..."
-	$(BENCH_BINARY) --sanity-check > /dev/null
+	@echo "Running bench/bench_bitcoin (one iteration sanity check, only high priority)..."
+	$(BENCH_BINARY) -sanity-check -priority-level=high > /dev/null
```
For the "Win64 native [vs2022]" CI task the same `-priority-level=high` has been added in #26481.
